os/arch/arm/src/amebad/amebad_timer_lowerhalf.c

```diff
@@ -0,0 +1,485 @@
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)lower;
+	uint32_t ticks;
+
+	/* Return the status bit */
+	status->flags = 0;
+	if (priv->started) {
+		status->flags |= TCFLAGS_ACTIVE;
+	}
+
+	if (priv->callback) {
+		status->flags |= TCFLAGS_HANDLER;
+	}
+
+	/* Return the actual timeout in microseconds */
+	status->timeout = priv->gpt_timeout;
+
+	/* Get the time remaining until the timer expires (in microseconds). */
+	ticks = gtimer_read_tick(&priv->obj);
+
+	if (priv->freerunmode) {	/* Free Run Mode */
+		status->timeleft = (uint64_t)((float)ticks * (1000000 / 32768) + (priv->freeruncount * 0xFFFFFFFF));
+	} else {					/* Alarm Mode */
+		status->timeleft = status->timeout - ticks * (1000000 / 32768);
+	}
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: amebad_gpt_settimeout
+ *
+ * Description:
+ *   Set a new timeout value (and reset the timer)
+ *
+ * Input Parameters:
+ *   lower   - A pointer the publicly visible representation of the
+ *             "lower-half" driver state structure.
+ *   timeout - The new timeout value in microseconds.
+ *
+ * Returned Values:
+ *   Zero on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+static int amebad_gpt_settimeout(struct timer_lowerhalf_s *lower, uint32_t timeout)
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)lower;
+	DEBUGASSERT(priv);
+
+	if (priv->started) {
+		return -EPERM;
+	}
+
+	gtimer_reload(&priv->obj, timeout);
+	priv->gpt_timeout = timeout;
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: amebad_gpt_setcallback
+ *
+ * Description:
+ *   Set the user provided timeout callback.
+ *
+ * Input Parameters:
+ *   lower    - A pointer the publicly visible representation of the
+ *              "lower-half" driver state structure.
+ *   callback - The new timer expiration function pointer. If this
+ *              function pointer is NULL, then the reset-on-expiration
+ *              behavior is restored,
+ *  arg       - Argument that will be provided in the callback
+ *
+ * Returned Values:
+ *   The previous timer expiration function pointer or NULL is there was
+ *   no previous function pointer.
+ *
+ ****************************************************************************/
+static void amebad_gpt_setcallback(struct timer_lowerhalf_s *lower, tccb_t callback, void *arg)
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)lower;
+	DEBUGASSERT(priv);
+
+	irqstate_t flags = irqsave();
+
+	/* Save the new callback */
+	priv->callback = callback;
+	priv->arg = arg;
+
+	/* If no callback is attached, the timer stop at the first interrupt */
+	if (callback != NULL && priv->started) {
+		priv->obj.handler = (void *)amebad_gpt_handler;
+		priv->obj.hid = priv;
+	} else {
```

Comment:
Could you please see below comments?
1. Mode is free-runcount, we can't register. In my opinion.
2. If mode is not alarm, we had better disable interrupt, right?
Also disable ISR.

Reply:
The freeruncount is not a standalone new mode, it is used to count the total time elapse in freerunmode. I will make changes and add description on it.

Please refer to this link to know the reason of using 'freerunmode', 'freeruncount' and the interrupt in Free Run mode, thanks!
https://github.com/Samsung/TizenRT/pull/4328#discussion_r450589393